### PR TITLE
test: add benchmark test suite for interceptor performance

### LINKED ISSUE #238: End to end and stress tests
We currently have mostly unit tests inside this repository, along with a few integration tests (like one for the interceptor coming in #206). I propose that we add more testing strategies to this project, including:

- **end to end**: a test that installs the addon to a Kubernetes cluster along with KEDA and any other dependencies, installs an app and an `HTTPScaledObject`, and runs a quick suite of requests to trigger scale-from-zero, scale up, and scale back down to zero. this test should check that the app and the interceptor both scale properly, and that the external scaler, operator, and interceptor's admin endpoints all report the expected things (NOTE: KEDA does this, and there is a prototype of something similar in #206)
- **stress**: a test that runs a large number of requests concurrently against an interceptor fleet, optionally for a large period of time. also possibly inspect internal admin endpoints along the way to ensure that metrics report as expected. the main purpose of this test is to ensure that, from the outside, requests get serviced as expected

### LINKED ISSUE #447: Provide automated load testing to benchmark performance of intercepting
Provide automated load testing to benchmark performance of intercepting.

We can use Artillery, Azure Load Testing, etc.

### PR DESCRIPTION
**Pull request by linkvt**
Measure interceptor performance in isolation (fixed replicas, no scaling) using oha as an in-cluster load generator.

We're using the ubuntu-latest runners with 4CPU and 16GB memory, IMO quite a lot for a decent load test. I made the config for the benchmark tests configurable via env vars in case we would need to change things for CI or local environments, as these thresholds are quite system specific.
Not really sure if all values need to be configurable but I chose to be consistent in the beginning and make everything here overwritable.

### Changes
- Add benchmark tests: throughput, max throughput, preloaded throughput and scale-from-zero cold-start latency
- Add oha helper for running load tests as Pods inside the cluster
- Add ScaleDeployment helper for cold-start test iterations
- Add GitHub Actions workflow (manual dispatch + PR&push to main) and Makefile targets
- Gate benchmark tests behind BENCHMARK=true so regular e2e runs skip them

### Checklist

- [x] Commits are signed with Developer Certificate of Origin ([DCO](https://developercertificate.org/))

Fixes #238 
Fixes #447 
Supersedes #1445

### CODE REVIEW

File: test/e2e/benchmark/many_routes_test.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+			t.Logf("creating %d filler InterceptorRoutes", benchCfg.RouteCount)
+			for i := range benchCfg.RouteCount {
+				f.CreateInterceptorRouteNoWait(
+					fmt.Sprintf("filler-%d-ir", i),
+					app,
+					h.IRWithHosts(fmt.Sprintf("filler-%d.example.com", i)),
+				)
+			}
+			time.Sleep(5 * time.Second)
+			t.Logf("all %d filler routes created", benchCfg.RouteCount)
```
[the remark this answers is not in the document]
Comment (linkvt) [reply test/e2e/benchmark/many_routes_test.go:37]: NO takes forever

File: .github/workflows/benchmark-tests.yaml [hunk 2]
```diff
@@ -0,0 +1,51 @@
+name: Benchmark
+on:
+  push:
+    branches: [main]
+  pull_request:
+    branches: [main]
+  workflow_dispatch:
```
[the remark this answers is not in the document]
Comment (linkvt) [reply .github/workflows/benchmark-tests.yaml:7]: ugh

File: test/e2e/benchmark/cold_start_test.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+			return ctx
+		}).
+		Assess("cold-start response times", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
+			f := h.NewFramework(ctx, t)
+
+			n := benchCfg.ColdStartSamples
+			durations := make([]time.Duration, 0, n)
+
+			for i := range n {
```
[the remark this answers is not in the document]
Comment (linkvt) [reply test/e2e/benchmark/cold_start_test.go:38]: how about not setting it to 0, great suggestion...

File: .github/workflows/benchmark-tests.yaml (hunk 2 above)
Comment (Fedosin) [inline .github/workflows/benchmark-tests.yaml:7]: Do we really need to run it on each push and pull_request to main? Maybe it's better to make this job optional?

**Review by Fedosin — COMMENTED** [review]
Do we really need to run it on each push and pull_request to main? Maybe it's better to make this job optional?

File: .github/workflows/benchmark-tests.yaml (hunk 2 above)
Comment (linkvt) [reply .github/workflows/benchmark-tests.yaml:7]: I think it makes sense to run them for every commit as part of our test suite as these benchmark tests have thresholds that would inform us in case of impacting changes.
The actual benchmark takes 9min, just 3min longer than other e2e tests. The e2e-setup setup took 44min for no apparent reason, it is much quicker usually as shown in the e2e tests that use the same make target.

When we get rid of the HTTPSO tests we will gain more CI time than we would spend with this single test.

Did a rebase besides that and switched to ARM for the benchmark test, lets see how this rerun behaves.

Comment (linkvt) [reply .github/workflows/benchmark-tests.yaml:7]: I forgot to set the `KO_DEFAULTPLATFORMS` env - e2e-setup should now be as fast as with the e2e tests.

**Review by Fedosin — APPROVED** [review]
